keras_cv/layers/preprocessing/clahe.py

```diff
@@ -0,0 +1,215 @@
+
+        numerator = cdf - cdf_min
+        denominator = tf.cast(tile_shape[0] * tile_shape[1], tf.float32) - cdf_min
+
+        cdf_normalized = tf.round(tf.math.divide_no_nan(numerator, denominator) * 255)
+        cdf_normalized = tf.cast(cdf_normalized, tf.int32)
+
+        # Reflection-pad the cdf functions so that we don't have to explicitly deal
+        # with corners/edge
+        cdf_padded = tf.pad(
+            cdf_normalized, [[0, 0], [1, 1], [1, 1], [0, 0]], mode="SYMMETRIC"
+        )
+
+        coords = tf.stack(
+            tf.meshgrid(
+                tf.range(tf.shape(image_padded)[0]),
+                tf.range(tf.shape(image_padded)[1]),
+                tf.range(tf.shape(image_padded)[2]),
+                indexing="ij",
+            )
+        )
+
+        y_coords = coords[0, :, :]
+        x_coords = coords[1, :, :]
+        z_coords = coords[2, :, :]
+
+        # Interpolation for efficient computation
+
+        half_tile_shape = tf.math.floordiv(tile_shape, 2)
+
+        nw_y_component = tf.math.floordiv(y_coords - half_tile_shape[0], tile_shape[0])
+        nw_x_component = tf.math.floordiv(x_coords - half_tile_shape[1], tile_shape[1])
+
+        # Need to correct negative values because negative-indexing for gather_nd ops
+        # not supported on all processors (cdf is padded to account for this)
+        nw_y_component = nw_y_component + 1
+        nw_x_component = nw_x_component + 1
+
+        ne_y_component = nw_y_component
+        ne_x_component = nw_x_component + 1
+
+        sw_y_component = nw_y_component + 1
+        sw_x_component = nw_x_component
+
+        se_y_component = sw_y_component
+        se_x_component = sw_x_component + 1
+
+        def cdf_transform(x_comp, y_comp):
+            gatherable = tf.stack([image_padded, y_comp, x_comp, z_coords], axis=-1)
+            return tf.cast(tf.gather_nd(cdf_padded, gatherable), tf.float32)
+
+        nw_transformed = cdf_transform(nw_x_component, nw_y_component)
+        ne_transformed = cdf_transform(ne_x_component, ne_y_component)
+        sw_transformed = cdf_transform(sw_x_component, sw_y_component)
+        se_transformed = cdf_transform(se_x_component, se_y_component)
+
+        y = (y_coords - half_tile_shape[0]) % tile_shape[0]
+        y = tf.cast(tf.math.truediv(y, tile_shape[0]), tf.float32)
+        x = (x_coords - half_tile_shape[1]) % tile_shape[1]
+        x = tf.cast(tf.math.truediv(x, tile_shape[1]), tf.float32)
+
+        # Interpolate
+        interpolated = (y * (x * se_transformed + (1 - x) * sw_transformed)) + (
+            1 - y
+        ) * (x * ne_transformed + (1 - x) * nw_transformed)
+
+        # Return image to original size and dtype
+        interpolated = interpolated[
+            0 : original_2d_shape[0], 0 : original_2d_shape[1], :
+        ]
+        interpolated = tf.cast(tf.round(interpolated), original_dtype)
+
+        return interpolated
+
+    def augment_image(self, image, transformation=None):
```

Comment:
So, do we only support 3 channel images?

In medical imaging it is quite common to have many more than 3 channels.